File: packages/chat-typescript/src/identity.ts
export type IdentityRole = 'user' | 'assistant' | 'system'

export type UserIdentity = {
  role: 'user'
  id: string
  name: string
  metadata: Record<string, unknown>
}

export type AssistantIdentity = {
  role: 'assistant'
  id: string
  name: string
  metadata: Record<string, unknown>
}

export type SystemIdentity = {
  role: 'system'
  id: string
  name: string
  metadata: Record<string, unknown>
}

export type Identity = UserIdentity | AssistantIdentity | SystemIdentity

export type IdentityWire = {
  role: IdentityRole
  id: string
  name: string
  metadata: Record<string, unknown>
}

const IDENTITY_ROLES: ReadonlySet<IdentityRole> = new Set(['user', 'assistant', 'system'])

export function isIdentity(value: unknown): value is Identity {
  if (typeof value !== 'object' || value === null) return false
  const record = value as Record<string, unknown>
  if (typeof record.role !== 'string' || !IDENTITY_ROLES.has(record.role as IdentityRole)) {
    return false
  }
  if (typeof record.id !== 'string' || typeof record.name !== 'string') return false
  if (typeof record.metadata !== 'object' || record.metadata === null) return false
  return true
}

export function parseIdentity(raw: unknown): Identity {
  if (!isIdentity(raw)) {
    throw new Error(`invalid identity: ${JSON.stringify(raw)}`)
  }
  return raw
}

File: packages/chat-python/src/rfnry_chat_protocol/identity.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter


class _IdentityBase(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    name: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class UserIdentity(_IdentityBase):
    role: Literal["user"] = "user"


class AssistantIdentity(_IdentityBase):
    role: Literal["assistant"] = "assistant"


class SystemIdentity(_IdentityBase):
    role: Literal["system"] = "system"


Identity = UserIdentity | AssistantIdentity | SystemIdentity

_identity_adapter: TypeAdapter[Identity] = TypeAdapter(
    UserIdentity | AssistantIdentity | SystemIdentity,
)


def parse_identity(raw: dict[str, Any]) -> Identity:
    return _identity_adapter.validate_python(raw)
